Add additional data to MUMPS interface

PolvereUnipi opened the pull request:
This PR refactors the basic MUMPS interface already existing and adds a constructor taking AdditionalData that can be used to control MUMPS execution. 

Added a few tests to check the exposed functionalities such as Block Low-Rank factorization. 

We plan to add MPI support as soon as there is a consensus on the current interface. 

FYI @fdrmrc @luca-heltai

luca-heltai commented:
The indent script fails, Davide:

```
Run ./contrib/utilities/check_indentation.sh
  ./contrib/utilities/check_indentation.sh
  shell: /usr/bin/bash -e {0}
Invalid author 'PolvereUnipi' without firstname and lastname

For possible solutions, consult the following page:
      https://github.com/dealii/dealii/wiki/Commit-authorship
```

Would you mind making sure that your full name is associated with your GitHub account?

luca-heltai commented on include/deal.II/lac/sparse_direct.h:
```diff
@@ -452,99 +452,146 @@ class SparseDirectUMFPACK : public EnableObserverPointer
 
   /**
-   * Exception
+   * A function that returns the ICNTL integer array from MUMPS.
+   * The ICNTL array contains integer control parameters for the MUMPS solver.
+   * ICNTL array is originally a fortran array, it's 1-based.
+   * To select the correct index use a macro like this: #define ICNTL(I)
+   * icntl[(I)-1] In the MUMPS documentation there is the description of each
+   * parameter of the array.
```
Maybe rephrase this by saying "keep in mind that MUMPS is a fortran library, and the documentation refers to indices into this array starting from one rather than from zero"

luca-heltai commented on tests/mumps/mumps_03.cc:
```diff
@@ -0,0 +1,148 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2002 - 2013 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
```
This is no longer the LICENCE we should add. Please use the correct LICENCE header, for this and for all other files. Copy it from one of the other headers: apache with LLVM exception.

luca-heltai reviewed (changes requested):
Please, add also a changelog file.

luca-heltai commented:
/rebuild

bangerth commented on include/deal.II/lac/sparse_direct.h:
```diff
@@ -452,99 +452,146 @@ class SparseDirectUMFPACK : public EnableObserverPointer
   /**
-   * Exception
+   * A function that returns the ICNTL integer array from MUMPS.
+   * The ICNTL array contains integer control parameters for the MUMPS solver.
+   * ICNTL array is originally a fortran array, it's 1-based.
+   * To select the correct index use a macro like this: #define ICNTL(I)
+   * icntl[(I)-1] In the MUMPS documentation there is the description of each
+   * parameter of the array.
    */
-  DeclException0(ExcInitializeAlreadyCalled);
+  int *
+  get_icntl();
```
This is quite a specialized interface. Why would a user need to have access to this information?

If you absolutely need this function, would you mind adding a note that ownership of the pointed to array remains with the current class, rather than with the *caller* of this function?

bangerth commented on include/deal.II/lac/sparse_direct.h:
```diff
@@ -452,99 +452,146 @@ class SparseDirectUMFPACK : public EnableObserverPointer
+   * A function that returns the ICNTL integer array from MUMPS.
+   * The ICNTL array contains integer control parameters for the MUMPS solver.
+   * ICNTL array is originally a fortran array, it's 1-based.
+   * To select the correct index use a macro like this: #define ICNTL(I)
+   * icntl[(I)-1] In the MUMPS documentation there is the description of each
```
```suggestion
   * To select the correct index one can use a macro like this: `#define ICNTL(I)
   * icntl[(I)-1]`. In the MUMPS documentation there is the description of each
```

bangerth commented on include/deal.II/lac/sparse_direct.h:
```diff
@@ -452,99 +452,146 @@ class SparseDirectUMFPACK : public EnableObserverPointer
+  mutable DMUMPS_STRUC_C id;
+#endif // DEAL_II_WITH_MUMPS
+
+  double                     *a;
+  mutable std::vector<double> rhs;
+  int                        *irn;
+  int                        *jcn;
```
We should eventually update this to use `std::unique_ptr` instead of raw pointers.

bangerth commented on source/lac/sparse_direct.cc:
```diff
@@ -861,9 +861,40 @@ SparseDirectUMFPACK::n() const
+  id.par = 1;
+  id.sym = 0;
+
+  // Use MPI_COMM_WORLD as communicator
+  id.comm_fortran = -987654;
```
Could we make this a run-time parameter as well?

bangerth replied:
Not necessarily as part of this patch, but keep it in the back of your mind.

bangerth commented on source/lac/sparse_direct.cc:
```diff
@@ -1004,49 +1003,54 @@ SparseDirectMUMPS::initialize(const Matrix &matrix)
 
   // Start solver
   id.job = 3;
   dmumps_c(&id);
```
I continue to be tickled by the fact that the MUMPS folks chose a system by which their public interface consists of *exactly one* function. I also question their software design skills.

bangerth commented on source/lac/sparse_direct.cc:
```diff
@@ -1099,4 +1101,4 @@ InstantiateMUMPS(SparseMatrix<double>) InstantiateMUMPS(SparseMatrix<float>)
     InstantiateMUMPS(BlockSparseMatrix<float>)
 #endif
 
-      DEAL_II_NAMESPACE_CLOSE
+      DEAL_II_NAMESPACE_CLOSE
```
You do actually need the terminating newline.

PolvereUnipi commented:
I have addressed most of the required changes. We opened a followup PR to remove the usage of raw pointers (#18475)

luca-heltai reviewed (changes requested):
Mine are all minor comments. Let's go forward with this if @bangerth and @kronbichler agree. @fdrmrc is proposing a new PR that adds supports for PETSc and Trilinos matrices (using the MPI version of mumps).

One thing that we should address soon is a PETSc and Trilinos independent version of a Parallel matrix. We have a parallel vector, support for MPI, and support for parallel MatrixFree infrastructure without Trilinos or PETSc, but we don't have support for a parallel matrix based code independent on those two libraries. If we find the consensus, we could use this occasion to add something.

bangerth commented:
@PolvereUnipi Let me know when you've addressed all comments. 

@luca-heltai Let's discuss the question of a parallel matrix in a separate issue. I'm not convinced that's a good use of our time.

PolvereUnipi commented on include/deal.II/lac/sparse_direct.h:
```diff
@@ -452,99 +452,187 @@ class SparseDirectUMFPACK : public EnableObserverPointer
   template <class Matrix>
   void
-  initialize(const Matrix &matrix, const Vector<double> &rhs_vector);
+  initialize_matrix(const Matrix &matrix);
```
initialize_matrix is a private routine that gets called by the actual initialize. The comment of that function will be fixed in the next commit. It was tied to a previous implementation and i forgot to change it: initialize_matrix does not initialize any mumps instance, it's just used to actually pass the matrix to MUMPS inside the actual initialize (which should be consistent with other interfaces).

bangerth commented:
You'll need to address these things here:
```
diff --git a/include/deal.II/lac/sparse_direct.h b/include/deal.II/lac/sparse_direct.h
index 0072ffb48..f08112a[5](https://github.com/dealii/dealii/actions/runs/15143510681/job/42638343932?pr=18466#step:4:6)0 100644
--- a/include/deal.II/lac/sparse_direct.h
+++ b/include/deal.II/lac/sparse_direct.h
@@ -[6](https://github.com/dealii/dealii/actions/runs/15143510681/job/42638343932?pr=18466#step:4:7)35,4 +635,4 @@ private:
 
 DEAL_II_NAMESPACE_CLOSE
 
-#endif // dealii_sparse_direct_h
\ No newline at end of file
+#endif // dealii_sparse_direct_h
diff --git a/tests/mumps/mumps_04.cc b/tests/mumps/mumps_04.cc
index ec200045c..06c2b3d34 100644
--- a/tests/mumps/mumps_04.cc
+++ b/tests/mumps/mumps_04.cc
@@ -15[9](https://github.com/dealii/dealii/actions/runs/15143510681/job/42638343932?pr=18466#step:4:10),4 +159,4 @@ main(int argc, char **argv)
   test<1>();
   test<2>();
   test<3>();
-}
\ No newline at end of file
+}
Error: Process completed with exit code 1.
```

PolvereUnipi commented:
@bangerth @luca-heltai 

I should have addressed all the issues you pointed out. Unique pointers should be addressed in the follow up pull request by @fdrmrc .
MPI communicator as run time parameter should be also addressed in the next pull request about MUMPS with MPI.

bangerth commented on include/deal.II/lac/sparse_direct.h:
```diff
@@ -452,97 +452,202 @@ class SparseDirectUMFPACK : public EnableObserverPointer
+   * A function that returns the ICNTL integer array from MUMPS.
+   * The ICNTL array contains integer control parameters for the MUMPS solver.
+   * Keep in mind that MUMPS is a fortran library and the documentation refers
+   * to indices into this array into this array starting from one rather than
```
```suggestion
   * to indices into this array starting from one rather than
```

bangerth commented on include/deal.II/lac/sparse_direct.h:
```diff
@@ -452,97 +452,202 @@ class SparseDirectUMFPACK : public EnableObserverPointer
 
   /**
-   * Constructor
+   * The right-hand side vector. This is the right hand side of the system
```
```suggestion
   * The right-hand side vector. This is the right hand side of the system.
```

bangerth commented on include/deal.II/lac/sparse_direct.h:
```diff
@@ -452,97 +452,202 @@ class SparseDirectUMFPACK : public EnableObserverPointer
   /**
-   * This function initializes a MUMPS instance and computes the factorization
-   * of the system's matrix <tt>matrix</tt>.
+   * The number of non-zero entries in the matrix.
+   */
+  types::global_dof_index nz;
```
People generally write this as `nnz` (n_nonzeros). Separately, though, I can relatively easily imagine matrices where you have a few million rows, but more than 4B nonzeros (at least in parallel). Is this a variable you have to pass to MUMPS at some point? Otherwise, I think you want to use `std::size_t` for this variable, or perhaps `std::uint64_t`.

Perhaps seen a different way: It makes sense to say that `n` is of type `types::global_dof_index` because it represents the total number of DoFs. But the total number of nonzeros is unrelated to the number of DoFs, and so it makes sense to also expect that it would have a different type.

bangerth reviewed (commented):
Just a small number of final comments. When you've got these fixed, would you mind squashing everything into one commit? It is not necessary to carry around 9 commits to eternity :-)

PolvereUnipi commented:
@bangerth @luca-heltai 

I think i did everything you asked. Thank you for your patience and help.

bangerth commented:
/rebuild

bangerth commented:
Nicely done!
